runtime/src/leader_bank_status.rs

```diff
@@ -0,0 +1,167 @@
+use {
+    crate::bank::Bank,
+    solana_sdk::slot_history::Slot,
+    std::{
+        sync::{Arc, Condvar, Mutex, RwLock, Weak},
+        time::{Duration, Instant},
+    },
+};
+
+/// Tracks leader status of the validator node and notifies when:
+///     1. A leader slot begins
+///     2. A leader slot completes
+#[derive(Debug, Default)]
+pub struct LeaderBankStatus {
+    /// Current state of the system
+    status: Mutex<Status>,
+    /// Weak reference to the current bank
+    bank: RwLock<Option<(Slot, Weak<Bank>)>>,
+    /// CondVar to notify status changes and waiting
+    condvar: Condvar,
+}
+
+/// Leader status state machine for the validator:
+/// [Unininitialized] -> [InProgress] -> [Completed] --|
+///                          ^-------------------------|
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+enum Status {
+    /// Initial state, no bank, but also not completed yet
+    #[default]
+    Uninitialized,
+    /// Slot is in progress as leader
+    InProgress,
+    /// PoH has reached the end of the slot, and the next bank as leader is not available yet
+    Completed,
+}
```

Comment:
how about removing `Uninitialized`? fewer state, less bugs, fewer bits:

```
enum Status {
  // the leader bank isn't available yet (just after initialized or emptied after poh-completed bank)
  StandBy,
  // poh-initiated bank is available
  InProgress,
}
```

Reply:
Yeah that makes sense - it seems fine since right now we only ever wait for a *new* notification of completed instead of just being completed/standby.

Done: bea9edccc539d3cc70b78890eab8c47618bbb0e5